### DISCUSSION TITLE
updating formula, package, installed via brew that is built from HEAD, issues ❗️

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#1058] [Everyday usage] [answered]
**TL;DR**

why doesn't `brew reinstall` have a `--HEAD` switch/flag for it? and can brew be updated to support it for updating packages that are installed using the `--HEAD` flag **without** having to uninstall/remove the package?

---

sometimes i need to install a formula from source ie.

```
brew install fish --build-from-source --verbose
```

☝️ that will indeed install the formula i specified from the tarball referenced in the formula file ie. at the time of writing this that would be fish version 3.2.0

however i've run into some issues with the latest fish release `3.2.0`

- one being related to my git prompt
- another being related to copying and pasting text [learn more](https://github.com/fish-shell/fish-shell/issues/7770)

so with that i need to install fish from `HEAD`

so i run the below command

```
brew install fish --HEAD --verbose
```

that works fine if the formula/package is not installed but if it is already installed then will not work and brew will show the following,

```
Warning: fish HEAD-efcfec0 is already installed and up-to-date.
To reinstall HEAD, run:
  brew reinstall fish
```

so what i believe to be the next logical step is to run the `reinstall` subcommand

```shell
brew reinstall fish --HEAD --verbose
```

and i'm presented with this lovely error message

```
Error: invalid option: --HEAD
```

basically the `reinstall` command appears to not have a `--HEAD` switch for updating/reinstalling formula from _head_

and i end up having to remove/uninstall the formula and reinstall it.

```
brew remove fish
brew install fish --HEAD --verbose
```

that ☝️ two command process seems to be the only way i can update packages that i have installed from _git HEAD_ and requires a lot of manual intervention

can brew be updated to support the below command

```shell
brew reinstall fish --HEAD --verbose
```

**Comment by carlocab** [comment, marked as answer]
Try
```
brew reinstall -s fish
```

**Reply by ipatch** [reply]
looking at the output `-s` is an alias for `--build-from-source` and as i tried to describe in my question, _building  from source_ is not the same as building from HEAD, as building from HEAD builds from the tip of git master/main branch more often than not, and using `-s` flag uses the tarball bundled up with a release.

i hope that makes sense.

**Reply by carlocab** [reply]
I know. Just try it, please.

**Reply by ipatch** [reply]
seems to have done the job.

```
brew reinstall -s --verbose fish
```

thanks for the heads up.

**Reply by ipatch** [reply]
this really should be documented better as the `-s` flag for the `install` and `reinstall` are essentially doing different things 🤷‍♂️

**Reply by carlocab** [reply]
Alternatively:
```
brew upgrade --fetch-HEAD -s fish
```

PRs to improve the docs here are welcome.

**Reply by ipatch** [reply]
yeah unfortunately i'm just the messenger 🏇dude

**Reply by ipatch** [reply]
and

```
brew upgrade --fetch-HEAD -s fish
```

will yield

```
Warning: fish HEAD-efcfec0 already installed
```

**Reply by carlocab** [reply]
> and
> 
> ```
> brew upgrade --fetch-HEAD -s fish
> ```
> 
> will yield
> 
> ```
> Warning: fish HEAD-efcfec0 already installed
> ```

That's because you already have the latest commit from the master branch installed. Try it again when your installation is not outdated.

**Reply by Bo98** [reply]
`brew reinstall` should remember if you have chosen `--HEAD`, which is why it doesn't have a `--HEAD` flag. I've opened https://github.com/Homebrew/brew/pull/10865 to fix the handling of that.
